### PR TITLE
[SPARK-48030][SQL] SPJ: cache rowOrdering and structType for InternalRowComparableWrapper

### PR DESCRIPTION
**Pull request by advancedxy**
### What changes were proposed in this pull request?
Cache rowOrdering and structType for InternalRowComparableWrapper

### Why are the changes needed?
For performance improvement

### Does this PR introduce _any_ user-facing change?
NO

### How was this patch tested?
Added a new benchmark to verify the performance improvement

### Was this patch authored or co-authored using generative AI tooling?
NO

### CODE REVIEW

**Review by yabola — APPROVED** [review]
After a quick look, this looks good to me. In addition, is there a better `expireAfterAccess` configuration for `NonFateSharingCache`?

**Comment by yabola** [comment]
@advancedxy  It appears that this PR can enhance performance when there are a large number of partitions. Could you please share the test results from a real DatasourceV2 table?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/InternalRowComparableWrapper.scala [hunk 1]
```diff
@@ -53,6 +55,21 @@ class InternalRowComparableWrapper(val row: InternalRow, val dataTypes: Seq[Data
 }
 
 object InternalRowComparableWrapper {
+  private final val MAX_CACHE_ENTRIES = 1024
```
[the remark this answers is not in the document]
Comment (szehon-ho) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/InternalRowComparableWrapper.scala:58]: Makes sense to me to have a config, though not familiar with spark preference for these things.

**Review by szehon-ho — COMMENTED** [review]
Sounds like a good find and promising results on real table.

**Review by sunchao — APPROVED** [review]
LGTM. It would be nice to have something to configure this but I don't think it is super important. I feel the default value should be more than enough for most use cases? Similarly for `expireAfterAccess`.

**Comment by sunchao** [comment]
Thanks! merged to master.

**Pull request closed without merging**